src/java.base/share/classes/jdk/internal/lang/stable/StableFieldUpdaterGenerator.java

```diff
@@ -0,0 +1,347 @@
+/*
+ * Copyright (c) 2025, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.  Oracle designates this
+ * particular file as subject to the "Classpath" exception as provided
+ * by Oracle in the LICENSE file that accompanied this code.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package jdk.internal.lang.stable;
+
+import jdk.internal.invoke.MhUtil;
+import jdk.internal.vm.annotation.ForceInline;
+
+import java.lang.invoke.MethodHandle;
+import java.lang.invoke.MethodHandles;
+import java.lang.invoke.MethodType;
+import java.lang.invoke.VarHandle;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * Stable field updater code generator.
+ */
+final class StableFieldUpdaterGenerator {
+
+    private StableFieldUpdaterGenerator() {}
+
+    private static final MethodHandles.Lookup LOCAL_LOOKUP = MethodHandles.lookup();
+
+    record Shape(Class<?> varType, List<Class<?>> coordinateTypes){
+
+        static Shape of(VarHandle varHandle) {
+            final List<Class<?>> shapeCoordinateTypes = new ArrayList<>(varHandle.coordinateTypes().size());
+            for (Class<?> coordinate: varHandle.coordinateTypes()) {
+                shapeCoordinateTypes.add(toObjectIfReference(coordinate));
+            }
+            return new Shape(toObjectIfReference(varHandle.varType()), shapeCoordinateTypes);
```

Comment:
This should construct an unmodifiable list:
```suggestion
            final var coordinateTypes = varHandle.coordinateTypes();
            final var shapeCoordinateTypes = new Object[coordinateTypes.size()];
            for (int i = 0; i < shapeCoordinateTypes.length; i++) {
                shapeCoordinateTypes[i] = toObjectIfReference(coordinateTypes.get(i));
            }
            return new Shape(toObjectIfReference(varHandle.varType()),
                             SharedSecrets.getJavaUtilCollectionAccess().listFromTrustedArray(shapeCoordinateTypes));
```